jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt

```diff
@@ -0,0 +1,397 @@
+import java.util.concurrent.TimeUnit
+import java.util.function.Supplier
+import kotlin.math.abs
+
+internal class BandwidthAllocator<T : MediaSourceContainer>(
+    eventHandler: EventHandler,
+    /**
+     * Provide the current list of endpoints (in no particular order).
+     * TODO: Simplify to avoid the weird (and slow) flow involving `endpointsSupplier` and `sortedEndpointIds`.
+     */
+    private val endpointsSupplier: Supplier<List<T>>,
+    /**
+     * Whether bandwidth allocation should be constrained to the available bandwidth (when `true`), or assume
+     * infinite bandwidth (when `false`.
+     */
+    private val trustBwe: Supplier<Boolean>,
+    parentLogger: Logger,
+    private val diagnosticContext: DiagnosticContext,
+    private val clock: Clock
+) {
+    private val logger = createChildLogger(parentLogger)
+
+    /**
+     * The estimated available bandwidth in bits per second.
+     */
+    private var bweBps: Long = -1
+
+    /**
+     * Whether this bandwidth estimator has been expired. Once expired we stop periodic re-allocation.
+     */
+    private var expired = false
+
+    /**
+     * The "effective" constraints for an endpoint indicate the maximum resolution/fps that this
+     * [BandwidthAllocator] would allocate for this endpoint given enough bandwidth.
+     *
+     * They are the constraints signaled by the receiver, further reduced to 0 when the endpoint is "outside lastN".
+     *
+     * Effective constraints are used to signal to video senders to reduce their resolution to the minimum that
+     * satisfies all receivers.
+     *
+     * With the multi-stream support added, the mapping is stored on a per source name basis instead of an endpoint id.
+     *
+     * When an endpoint falls out of the last N, the constraints of all the sources of this endpoint are reduced to 0.
+     *
+     * TODO Update this description when the endpoint ID signaling is removed from the JVB.
+     */
+    private var effectiveConstraints = emptyMap<String, VideoConstraints>()
+    private val eventEmitter: EventEmitter<EventHandler> = SyncEventEmitter()
+
+    /**
+     * The allocations settings signalled by the receiver.
+     */
+    private var allocationSettings =
+        AllocationSettings(defaultConstraints = VideoConstraints(BitrateControllerConfig.config.thumbnailMaxHeightPx()))
+
+    /**
+     * The last time [BandwidthAllocator.update] was called.
+     */
+    private var lastUpdateTime: Instant
+
+    /**
+     * The result of the bitrate control algorithm, the last time it ran.
+     */
+    var allocation = BandwidthAllocation(emptySet())
+        private set
+
+    /**
+     * The task scheduled to call [.update].
+     */
+    private var updateTask: ScheduledFuture<*>? = null
+
+    init {
+        eventEmitter.addHandler(eventHandler)
+        // Don't trigger an update immediately, the settings might not have been configured.
+        lastUpdateTime = clock.instant()
```

Comment:
Is this in `init` to make sure it happens after `addHandler` and before `rescheduleUpdate`?

Reply:
No, that's just how IDEA translated it to kt. I cleaned it up